zenoh-ts/src/query.ts

```diff
@@ -0,0 +1,574 @@
+
+  /**
+  * Returns true if properties does not contain anything.
+  * @returns void
+  */
+  is_empty(): boolean {
+    return (this._params.size == 0);
+  }
+
+  /**
+   * checks if parameters contains key
+   * @returns boolean
+   */
+  contains_key(key: string): boolean {
+    return this._params.has(key)
+  }
+
+  /**
+   * gets value with associated key, returning undefined if key does not exist
+   * @returns string | undefined
+   */
+  get(key: string): string | undefined {
+    return this._params.get(key)
+  }
+
+  /**
+   * Inserts new key,value pair into parameter
+   * @returns void
+   */
+  insert(key: string, value: string) {
+    return this._params.set(key, value);
+  }
+
+  /**
+   * extends this Parameters with the value of other parameters, overwriting `this` if keys match.  
+   * @returns void
+   */
+  extend(other: IntoParameters) {
+    let other_params = Parameters.new(other);
+    for (let [key, value] of other_params._params) {
+      this._params.set(key, value)
+    }
+  }
+
+  /**
+   * returns the string representation of the 
+   * @returns void
+   */
+  toString(): string {
+    let output_string = "";
+    for (let [key, value] of this._params) {
+      output_string += key + "=" + value + ";"
+    }
+    output_string = output_string.substring(0, output_string.length - 1);
+
+    return output_string;
+  }
+
+  static new(p: IntoParameters): Parameters {
+    if (p instanceof Parameters) {
+      return p
+    } else if (p instanceof Map) {
+      return new Parameters(p);
+    } else {
+      const params = new Map<string, string>();
+      if (p.length != 0) {
+        for (const pair of p.split(";") || []) {
+          const [key, value] = pair.split("=");
+          if (key != undefined && value != undefined) {
+            params.set(key, value);
+          }
+        }
+      }
+      return new Parameters(params);
+    }
+  }
+}
+
+
+/**
+ * ReplyError returned from a `get` on a session
+ * 
+ */
+export class ReplyError {
+  private _payload: ZBytes;
+  private _encoding: Encoding;
+
+  /**
+   * Payload of Error Reply
+   * @returns ZBytes
+   */
+  payload(): ZBytes {
+    return this._payload;
+  }
+
+  /**
+   * Encoding of Error Reply
+   * @returns ZBytes
+   */
+  encoding(): Encoding {
+    return this._encoding;
+  }
+
+  private constructor(payload: ZBytes, encoding: Encoding) {
+    this._encoding = encoding;
+    this._payload = payload;
+  }
+
+  /**
+    * ReplyError gets created by the reply of a `get` on a session
+    * @hidden
+    */
+  static new(reply_err_ws: ReplyErrorWS): ReplyError {
+    let zbytes = ZBytes.new(reply_err_ws.payload);
+    let encoding = Encoding.from_str(reply_err_ws.encoding);
+    return new ReplyError(zbytes, encoding);
+  }
+}
+
+/**
+ * Reply object from a zenoh `get`
+ */
+export class Reply {
+  private _result: Sample | ReplyError;
+
+  /**
+   * Payload of Error Reply
+   * @returns Sample or ReplyError 
+   */
+  result(): Sample | ReplyError {
```

Comment:
I think this ought to use an exception instead.

Reply:
I think use of an exception here may make the API inconsistent with Rust.

Reply:
You have a point, but exceptions are the idiomatic method of error handling in JS/TS. In Rust exceptions are frowned upon in favor of returning errors.

For instance, zenoh-python also uses exceptions:

https://github.com/eclipse-zenoh/zenoh-python/blob/d686a576ce760f0ff3dab8832fe22a6ff390043c/examples/z_get.py#L24-L31